### PR TITLE
remove the limit of 1024 records per query (#57671)

### LINKED ISSUE #57671: na_ontap_gather_facts limited to 1024 records
##### SUMMARY

The na_ontap_gather_facts module has a hardcoded limit of 1024 entries ('max-records': '1024') on queries to the ONTAP API. It does not use next-tag to query for additional entries, which means the facts will not contain all entries (for example volumes) on larger clusters.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

na_ontap_gather_facts

##### ANSIBLE VERSION

```paste below
2.7.6
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

ONTAP 9.3P12

##### STEPS TO REPRODUCE

Run na_ontap_gather_facts on a cluster with more than 1024 volumes and compare the facts cache with the volumes on the system.

```yaml

```

##### EXPECTED RESULTS

The module should do additional queries to the API using the next-tag parameter until all entries have been read.

##### ACTUAL RESULTS

Only 1024 out of 1422 volumes included in the facts.

```paste below

```

### PR DESCRIPTION
**Pull request by dprts**
##### SUMMARY
Fixes #57671
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
na_ontap_gather_facts

##### ADDITIONAL INFORMATION
optional parameter added:
- max_records
        Limits maximum amount of records returned per query. It must be between 1 and 4294967295.
        [Default: 1024]
        type: str
        version_added: 2.9

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py [hunk 1]
```diff
@@ -46,6 +46,13 @@
             - use "help" to get a list of supported facts for your system.
         default: "all"
         version_added: 2.8
+    max_records:
+        description:
+            - Limits maximum amount of records returned per query.
+                It must be between 1 and 4294967295.
+        default: "1024"
+        type: str
```
Comment (lonico) [inline lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py]: Why not int ?

I’d think it’s more natural to use int for a number and to convert at the last minute to match ZAPI expectations.   This is even more useful as we will transition to the new REST APIs as they are being supported in ONTAP 9.6 and newer releases.

Comment (dprts) [reply lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py]: ok, i'll look into this

Comment (dprts) [reply lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py]: I agree and have changed it to int ;)

File: lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py [hunk 2]
```diff
@@ -46,6 +46,13 @@
             - use "help" to get a list of supported facts for your system.
         default: "all"
         version_added: 2.8
+    max_records:
+        description:
+            - Limits maximum amount of records returned per query.
+                It must be between 1 and 4294967295.
+        default: "1024"
```
Comment (lonico) [inline lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py]: I wouldn’t use quotes

File: test/units/modules/storage/netapp/test_na_ontap_gather_facts.py [hunk 3]
```diff
@@ -123,8 +123,15 @@ def get_gather_facts_mock_object(self, kind=None):
         """
         Helper method to return an na_ontap_gather_facts object
         """
+
+        argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        argument_spec.update(dict(
+            state=dict(default='info', choices=['info']),
+            gather_subset=dict(default=['all'], type='list'),
+            max_records=dict(default='1024', type='int'),
```
Comment (lonico) [inline test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: Quotes?

**Review by lonico — APPROVED** [review]
shipit

File: lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py (hunk 2 above)
Comment (dprts) [reply lib/ansible/modules/storage/netapp/na_ontap_gather_facts.py]: right, this is misleading. i will correct it here

File: test/units/modules/storage/netapp/test_na_ontap_gather_facts.py (hunk 3 above)
Comment (dprts) [reply test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: ... and here

Comment (carchi8py) [reply test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: Since its an int i think default has to be 1024 with out quotes.

**Review by carchi8py — COMMENTED** [review]
Thanks @dprts for making that change so quickly!

File: test/units/modules/storage/netapp/test_na_ontap_gather_facts.py (hunk 3 above)
Comment (dprts) [reply test/units/modules/storage/netapp/test_na_ontap_gather_facts.py]: yes, sorry, i forgot to add testing module to this commit.

**Review by lonico — APPROVED** [review]
shipit

**Comment by carchi8py** [comment]
It looks like this didn't make it in to 2.9. I'll see if we can bring this in to the collection for it first release.

**Pull request closed without merging**